Stargate v2 readme updates (#2055)
stargate/stargate: README.md

> It's built with extensibility as a first-class citizen and makes it easy to use a database for any application workload

Reviewer: ```suggestion
CQL is built with extensibility as a first-class citizen, which makes it easy to use a database for any application workload
```

This project enables customization of all aspects of data access and has modules for authentication, APIs, request handling / routing, and persistence backends.
The current form is specific to the Apache Cassandra (C*) backend but there's no bounds to the databases or APIs that this framework can support.

As shown in the figure below, Stargate is often deployed behind a load balancer or proxy and exposes multiple endpoints to client applications, including HTTP APIs, gRPC, and the Cassandra Query Language (CQL). Stargate sits in front of a Cassandra cluster which is used as the storage backend. 

![image](assets/stargate-arch-high-level.png#center)

Stargate consists of the following components, which we introduce briefly here with links to the corresponding modules in this monorepo.

### API Services
These are independently scalable microservices which various APIs, typically HTTP based. These modules can be found under the [apis](apis) directory: 

- [sgv2-restapi](apis/sgv2-restapi): API implementation for exposing Cassandra data over REST
- [sgv2-graphqlapi](apis/sgv2-graphqlapi): API implementation for exposing Cassandra data over GraphQL

> - [sgv2-docsapi](apis/sgv2-graphqlapi): API implementation for exposing Cassandra data over GraphQL

Reviewer: ```suggestion
- [sgv2-docsapi](apis/sgv2-docsapi): API implementation for exposing Cassandra data over a Document API
```

### Coordinator Node
Coordinator nodes participate as non-data storing nodes in the backing Cassandra cluster, which enables them to read and write data more efficiently. Stargate Coordinator nodes can also be scaled independently. Coordinator nodes expose gRPC and CQL interfaces for fast access by client applications. The following are the key modules comprising the coordinator and its exposed interfaces:

- [core](core): Common classes used throughout the other coordinator modules
- [cql](cql): API implementation for the Cassandra Query Language
- [grpc](grpc): fast CQL over gRPC implementation (HTTP-based interface equivalent to CQL performance)
- [bridge](bridge): gRPC-based interface used by API services
- [health-checker](core): HTTP endpoints useful for health checking coordinator nodes
- [metrics-jersey](core): metrics collection for the coordinator node and its exposed interfaces
- [stargate-starter](stargate-starter): the main Java application used to start the coordinator via the `starctl` script

#### Persistence Services

> Stargate coordinator nodes support a pluggable approach for implementing the coordination layer to execute requests passed by API services and other interfaces to underlying data storage instances. Persistence service implementations are responsible handling and converting requests to db queries, dispatching to persistence, returning and serving response

Reviewer: ```suggestion
Stargate coordinator nodes support a pluggable approach for implementing the coordination layer to execute requests passed by API services and other interfaces to underlying data storage instances. Persistence service implementations are responsible handling and converting requests to database queries, dispatching to persistence, and returning and serving responses.
```

> Stargate coordinator nodes support a pluggable approach for implementing the coordination layer to execute requests passed by API services and other interfaces to underlying data storage instances. Persistence service implementations are responsible handling and converting requests to database queries, dispatching to persistence, and returning and serving responses.

Reviewer 2: Could also mention that they expose unified interface to hide differences between various Cassandra versions, so that API services operate against "single" backend, essentially (although TBH with v2 I guess this is more due to Bridge exposing unified API).

Author: added detail

> - [persistence-api](persistence-api): Interface for working with persistence services.  as   

Reviewer: ```suggestion
- [persistence-api](persistence-api): Interface for working with persistence services. 
```

> - [persistence-common](persistence-common): Utilities shared by the persistence services

Reviewer: ```suggestion
- [persistence-common](persistence-common): Utilities shared by the persistence services.
```

> - [persistence-cassandra-3.11](persistence-cassandra-3.11): Joins C* 3.11 cluster as coordinator-only node (does not store data),

Reviewer: ```suggestion
- [persistence-cassandra-3.11](persistence-cassandra-3.11): Joins Cassandra 3.11 cluster as coordinator-only node (does not store data),
```

> mocks C* system tables for native driver integration,

Reviewer: As far as I know, `C*` is a pretty DS-internal designation. I would spell out C* as Cassandra in every instance of the use of this abbreviation (?).

Author: I think C* pretty generally used and understood in the community, and it is introduced at the top of the file (line 28).

executes requests with C* storage nodes using C* QueryHandler/QueryProcessor,
converts internal C* objects and ResultSets to Stargate Datastore objects.
- [persistence-cassandra-4.0](persistence-cassandra-4.0): (same as above but for Cassandra 4.0)
- [persistence-dse-6.8](persistence-dse-6.8): (same as above but for DataStax Enterprise 6.8)

#### Authentication Services
Stargate coordinator nodes also support a pluggable authentication approach.

> - [authnz](authnz): Interface for working with auth providers

Reviewer: I once got chewed out about the use of `authz` - isn't that authorization, not authentication? You might want to make this statement clearer about that distinction.

Author: the module name is `authnz` which is an abbreviation for both. I'll spell out auth providers so it is more clear

Author: I actually changed the title and first sentence of this section

> - [auth-api](auth-api): REST service for generating auth tokens

Reviewer: You used periods at the end of the descriptions in the last table. To be consistent, you should add them here. Or take them away in all cases. Consistency is the important point.

Author: will make consistent

- [auth-table-based-service](auth-table-based-service): Service to store tokens in the database
- [auth-jtw-based-service](auth-jwt-based-service): Service to authenticate using externally generated JSON Web Tokens (JWTs)

#### Coordinator Node Testing
The following modules provide support for testing:

- [testing](testing): Integration test suite for the coordinator node modules
- [persistence-test](persistence-test): Common utilities for testing persistence services

Instructions for running and extending the test suite can be found in the [developer guide](DEV_GUIDE.md).

## Repositories

Here is an overview of the key repositories in the Stargate GitHub organization:

- [stargate/stargate](https://github.com/stargate/stargate): This repository is the primary entry point to the project. It is a monorepo containing all of the Stargate modules.
- [stargate/docs](https://github.com/stargate/docs): This repository contains the user docs hosted on [stargate.io](https://stargate.io)

> - [stargate/website](https://github.com/stargate/website): This repository contains the code for the website hosted on [stargate.io](https://stargate.io)

Reviewer: I think it would be useful to add stargate/docker-images, and maybe the various grpc client repos.

Reviewer 2: @polandll `docker-images` will go away as it is only used by Stargate V1. I think.

Author: Tatu is correct about `docker-images`. I can add a general statement about gRPC repos rather than an exhaustive list.
